prestoadmin/server.py

```diff
@@ -82,11 +85,218 @@ def new_sysnode_processor(node_info_rows):
+    return response_status == 200
+
+
+def _get_content_length(url_response):
+    try:
+        headers = url_response.info()
+        return int(headers['Content-Length'])
+    except (KeyError, ValueError):
+        return None
+
+
+def _get_download_file_name(url_response, version=None):
+    try:
+        headers = url_response.info()
+        content_disposition = headers['Content-Disposition']
+        values, params = cgi.parse_header(content_disposition)
+        return params['filename']
+    except KeyError:
+        if not version:
+            return DEFAULT_RPM_NAME
+        else:
+            return 'presto-server-rpm-' + version + '.rpm'
+
+
+def _print_download_status(bytes_read, total_bytes):
+    percent = float(bytes_read) / total_bytes
+    percent = round(percent * 100, 2)
+    print 'Downloaded %d of %d bytes. (%0.2f%%)' % \
+          (bytes_read, total_bytes, percent)
+
+
+def _find_downloaded_rpm(download_file_name):
+    download_file_path = os.path.join(DOWNLOAD_DIRECTORY, download_file_name)
+
+    sys.stderr = open(os.devnull, 'w')
+    try:
+        package.check_if_valid_rpm(download_file_path)
+        print 'Found rpm at: %s' % download_file_path
+    except SystemExit as system_exit:
+        if 'Corrupted RPM' in system_exit.message:
+            print 'Removing corrupted rpm at: %s' % download_file_path
+            try:
+                os.remove(download_file_path)
+            except OSError:
+                pass
+        return None
+    finally:
+        sys.stderr = sys.__stderr__
+
+    return download_file_path
+
+
+def _download_rpm(url_response, download_file_path):
+    content_length = _get_content_length(url_response)
+    print 'Downloading rpm from %s\n' \
+          'This can take a few minutes.' % url_response.geturl()
+
+    with open(download_file_path, 'wb') as local_file:
+        bytes_read = 0
+        block_size = 16 * 1024 * 1024
+        while True:
+            download_buffer = url_response.read(block_size)
+            if not download_buffer:
+                break
+            bytes_read += len(download_buffer)
+            local_file.write(download_buffer)
+            if content_length:
+                _print_download_status(bytes_read, content_length)
+        print "Downloaded %d bytes." % bytes_read
+
+    print 'Rpm downloaded to: %s' % download_file_path
+    return download_file_path
+
+
+def _find_or_download_rpm(url, version=None):
+    """
+    Args:
+        url:      The url of the presto rpm to be downloaded.
+        version:  An optional version number.
+                  If the server doesn't respond with the file name that is being
+                  requested, this allows the downloaded file to have the correct
+                  version attached to its name (presto-server-rpm-'version'.rpm)
+                  rather than the default name
+
+    If downloading the presto rpm at the given url would overwrite an existing rpm,
```

Comment:
Also: This is totally user-hostile for the case of a partial download or corrupted file.